mautic/mautic
# TPROD-260 Marketplace: allow installing/removing plugins from within Mautic

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 4.1
| New feature/enhancement? (use the a.x branch)      | yes
| Deprecations?                          | no
| BC breaks? (use the c.x branch)        | no
| Automated tests included?              | yes
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation-new/pull/10
| Issue(s) addressed                     | Fixes #... 



#### Description:

In Mautic 4.0.0, the Marketplace was added as read-only beta. This PR will add the functionality to either install or remove plugins from Mautic's UI or CLI 🎉 

Some key info:
- We add `composer/composer` as a dependency to Mautic itself. This way, we can control the version of Composer we use. Next to that, some hosting environments have different PHP versions in the CLI and webserver, so by running Composer in-process we can be sure that we run the same PHP version that Mautic runs in
- Bumped `friendsofphp/php-cs-fixer` to 2.19.2 as that version supports `composer/xdebug-handler` 2.0.0 and higher, which is needed by the `composer/composer` package. Proper PR which includes code style updates: https://github.com/mautic/mautic/pull/10519
- Explicitly defined Mautic's version in `app/composer.json`. This is needed for local development environments to work correctly, otherwise it's not possible to install plugins using Composer (will add some additional info/explanation later).

You can already run `php bin/console mautic:marketplace:install shinde-rahul/helloworld` which will install the helloworld plugin into the `plugins` folder.

To do:
- [x] Automatically clear cache after plugin installation/removal
- [x] Add UI functionality for installing plugins (we can use our new `Mautic\MarketplaceBundle\Service\Composer` service for that)
- [x] Create a `composer.json` that we can include in Mautic's installation ZIP (it's currently excluded in the build process). We need to decide if we want to include dev dependencies (php-cs-fixer, PHPSTAN, etc.) in it as well
- [x] Create developer docs (https://github.com/mautic/developer-documentation-new/pull/10)
- [x] Create end-user docs (ping Favour for that)
- [x] Set up allowlist and parse it in the Marketplace (https://github.com/mautic/marketplace-allowlist)
- [ ] Ensure that Mautic installs stable plugins only (or provide a dropdown to select version) - IIRC from testing, our current Composer config is set up in such way that doesn't prefer stable packages so will install dev/beta versions of plugins which we probably don't want. But will need to re-test this
- [x] (see https://github.com/mautic/mautic/pull/10518#issuecomment-1024905148) - Bundle a `composer.phar` with Mautic - updating Mautic through the UI/CLI has changed slightly in this PR. We still overwrite all files from the update ZIP file, except for the vendor folder and composer.json/composer.lock. Otherwise we would just overwrite all of the user’s installed plugins on each Mautic update :sweat_smile: In the PR I’m already adding `composer/composer` as a dependency to Mautic itself. However, this becomes problematic when it tries to update `composer/composer` itself and things break horribly. Hence the suggestion for adding `composer.phar` as a “standalone” executable
- [x] Add the new plugin events that Alan shared in https://github.com/mautic/mautic/pull/10518#issuecomment-990419709
- [ ] Officially deprecate updating Mautic through the UI (update the docs accordingly), and remove it in M5, in favor of `composer update`.
- [ ] Update release leader doc to bump the Mautic (`mautic/core-lib`) version on every Mautic update. Otherwise plugins can’t be installed if they depend on a minimum `mautic/core-lib` version
- [ ] Fill the allowlist with plugins that we tested manually with Mautic https://github.com/mautic/marketplace-allowlist
- [ ] Create guidelines for developers submitting new plugins to the allowlist
- [x] Update the old developer docs portal for the Marketplace 



#### Steps to test this PR:


1. Load up this PR - you can either do that in Gitpod (see below) or locally through DDEV. **Make sure you use `https://MAUTIC_URL/index_dev.php` as the production assets haven't been updated yet!**
2. Go into Mautic's Configuration > System Settings > Update Settings and enable `Update Mautic through Composer [BETA]`:
<img width="1158" alt="image" src="https://user-images.githubusercontent.com/17739158/153424152-da8e2120-6564-4a8c-bb3e-48973d6bd523.png">
3. Open the Marketplace. You should be able to see the "Helloworld bundle" and install it:
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/17739158/153424481-52b92c1c-e2a3-4652-b8a2-bc9e3e81f11a.png">
4. Please also try to remove the plugin afterwards to see if that process also works correctly




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10518"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
